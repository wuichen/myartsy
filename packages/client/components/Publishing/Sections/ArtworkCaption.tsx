import { color as Color, Sans, Serif } from "@artsy/palette"
import { ErrorBoundary } from "Components/ErrorBoundary"
import { pMedia } from "Components/Helpers"
import { ArticleLayout, SectionLayout } from "Components/Publishing/Typings"
import { Truncator } from "Components/Truncator"
import _ from "lodash"
import React from "react"
import styled from "styled-components"

interface ArtworkCaptionProps extends React.HTMLProps<HTMLDivElement> {
  artwork: any
  color?: string
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
  linked?: boolean
  isFullscreenCaption?: boolean
}

export class ArtworkCaption extends React.Component<ArtworkCaptionProps> {
  static defaultProps = {
    color: Color("black30"),
  }

  joinParts(children, key, delimiter = ", ") {
    const compacted = _.compact(children)

    if (compacted.length) {
      const reduced = compacted.reduce((prev, curr, i) => {
        return [
          prev,
          <span key={`joinParts-${key}-${i}`}>{delimiter}</span>,
          curr,
        ]
      })
      return reduced
    } else {
      return []
    }
  }

  joinArtistNames(names, delimiter = ", ") {
    if (names.length === 0) {
      return []
    }

    return names.slice(1).reduce(
      (prev, curr, i) => {
        return prev.concat([
          <span key={`joinArtistNames-${i}`}>{delimiter}</span>,
          curr,
        ])
      },
      [names[0]]
    )
  }

  renderArtists() {
    const {
      artwork: { artist, artists },
    } = this.props

    // Multiple artists
    if (artists && artists.length > 1) {
      const names = artists.map((a, i) => {
        const artistName = this.renderArtistName(a, `renderArtists-${i}`)
        return artistName
      })

      const joinedNames = this.joinArtistNames(names)
      return joinedNames

      // Single artist
    } else if (artist) {
      const artistName = this.renderArtistName(artist, "renderArtists-single")
      return artistName
    }
  }

  renderArtistName(artist, key: string) {
    const { linked } = this.props
    const { name, slug } = artist
    const createTextLink = linked && slug

    if (createTextLink) {
      const href = `/artist/${slug}`

      return (
        <ArtistName key={`renderArtistName-${key}`}>
          <a href={href}>{name}</a>
        </ArtistName>
      )
    } else {
      return (
        <span key={`renderArtistName-${key}`} className="name">
          {name}
        </span>
      )
    }
  }

  renderTitleDate() {
    const children = [this.renderTitle(), this.renderDate()]
    const titleDate = this.joinParts(children, "renderTitleDate")
    return titleDate
  }

  renderTitle() {
    const {
      artwork: { slug, title },
      linked,
    } = this.props

    if (title) {
      if (linked) {
        const href = `/artwork/${slug}`

        return (
          <span key="renderTitle" className="title">
            <a href={href}>{title}</a>
          </span>
        )
      } else {
        return (
          <span key="renderTitle" className="title">
            {title}
          </span>
        )
      }
    }
  }

  renderDate() {
    const {
      artwork: { date },
    } = this.props

    if (date && date.length) {
      return (
        <span key="renderDate" className="date">
          {date}
        </span>
      )
    }
  }

  renderPartner() {
    const {
      artwork: {
        partner: { name, slug },
      },
      linked,
    } = this.props

    if (name) {
      const createTextLink = Boolean(linked && slug)

      if (createTextLink) {
        return (
          <a key="renderPartner" href={`/${slug}`}>
            {name}
          </a>
        )
      } else {
        return name
      }
    }
  }

  renderCredit() {
    const {
      artwork: { credit },
    } = this.props

    if (credit && credit.length) {
      return (
        <span key="renderCredit" className="credit">
          {credit}
        </span>
      )
    }
  }

  renderPartnerCredit = () => {
    const children = [this.renderPartner(), this.renderCredit()]

    const joined = this.joinParts(children, "renderPartnerCredit", ". ")
    return joined
  }

  renderFullscreenCaption = () => {
    return (
      <StyledFullscreenCaption size={["3", "4"]} weight="medium">
        <Line>
          <ArtistNames>{this.renderArtists()}</ArtistNames>
        </Line>
        <div>
          <Line>{this.renderTitleDate()}</Line>
          <Line>{this.renderPartnerCredit()}</Line>
        </div>
      </StyledFullscreenCaption>
    )
  }

  renderClassicCaption = () => {
    return (
      <StyledClassicCaption
        color={Color("black60")}
        size="2"
        className="display-artwork__caption"
      >
        <Truncator>
          <ArtistNames>{this.renderArtists()}</ArtistNames>
          {this.renderTitleDate()}
          {". "}
          {this.renderPartner()}
        </Truncator>
      </StyledClassicCaption>
    )
  }

  renderEditorialCaption = () => {
    const { color, layout, sectionLayout } = this.props

    return (
      <StyledArtworkCaption
        size="3"
        color={color}
        layout={layout}
        sectionLayout={sectionLayout}
      >
        <ArtistNames>{this.renderArtists()}</ArtistNames>
        <div>
          <Truncator>{this.renderTitleDate()}</Truncator>
          <Truncator>{this.renderPartnerCredit()}</Truncator>
        </div>
      </StyledArtworkCaption>
    )
  }

  render() {
    const { layout, isFullscreenCaption } = this.props

    return (
      <ErrorBoundary>
        <div>
          {isFullscreenCaption
            ? this.renderFullscreenCaption()
            : layout === "classic"
            ? this.renderClassicCaption()
            : this.renderEditorialCaption()}
        </div>
      </ErrorBoundary>
    )
  }
}

const ArtistNames = styled.span`
  margin-right: 30px;
`

const ArtistName = styled.span`
  white-space: nowrap;
`

export const StyledArtworkCaption = styled(Sans)<{
  color?: string
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
}>`
  padding: ${props => (props.sectionLayout === "fillwidth" ? "0 10px;" : "0;")};
  margin-top: 10px;
  display: flex;

  a {
    color: ${props => props.color};
    text-decoration: none;
  }

  .title,
  .title a {
    font-style: italic;
  }

  ${pMedia.xs`
    padding: 0 10px;
  `};
`

const StyledClassicCaption = styled(Serif)<{
  className?: string
  color?: string
}>`
  margin-top: 10px;
  display: block;

  a {
    color: ${props => props.color};
    text-decoration: none;
  }

  ${ArtistNames} {
    margin-right: 0;
    font-weight: bold;

    &::after {
      content: ", ";
    }
  }

  .title {
    font-style: italic;
  }
`

const StyledFullscreenCaption = styled(Sans)`
  display: flex;

  a {
    color: black;
    text-decoration: none;
  }

  .title,
  .title a {
    font-style: italic;
  }

  ${pMedia.sm`
    flex-direction: column;
  `};
`

const Line = styled.div`
  ${pMedia.sm`
    &.artist-name {
      margin-bottom: 5px;
    }
  `};
`
