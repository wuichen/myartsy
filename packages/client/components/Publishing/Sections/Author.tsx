import { Box, Flex, Sans, space } from "@artsy/palette"
import { pMedia } from "Components/Helpers"
import Icon from "Components/Icon"
import React from "react"
import Markdown from "react-markdown"
import styled from "styled-components"
import { resize } from "Utils/resizer"

interface AuthorProps {
  author: any
  color?: string
}

export const Author: React.SFC<AuthorProps> = ({ author, color = "black" }) => {
  const profileImage = author.image_url ? (
    <ProfileImage mr={20} src={resize(author.image_url, { width: 200 })} />
  ) : (
    false
  )
  return (
    <AuthorContainer color={color} mb={20} alignItems="center">
      {profileImage}
      <Sans size="4" weight="medium">
        <AuthorInfo>
          {author.bio && author.bio.length ? (
            <Markdown
              source={author.bio}
              disallowedTypes={["Paragraph"]}
              unwrapDisallowed
              containerTagName="span"
            />
          ) : (
            <div>{author.name}</div>
          )}
        </AuthorInfo>

        {author.twitter_handle && author.twitter_handle.length ? (
          <span>
            <TwitterHandle href={`http://twitter.com/${author.twitter_handle}`}>
              <Icon name="twitter" color={color} />
              {`@${author.twitter_handle}`}
            </TwitterHandle>
          </span>
        ) : (
          false
        )}
      </Sans>
    </AuthorContainer>
  )
}

const ProfileImage = styled(Box)<{ src?: string }>`
  min-width: 60px;
  min-height: 60px;
  border-radius: 50%;
  background: url(${props => props.src || ""}) no-repeat center center;
  background-size: cover;

  ${pMedia.xs`
    min-width: 40px;
    min-height: 40px;
  `};
`

const AuthorContainer = styled(Flex)<{ color: string }>`
  color: ${props => props.color};

  a {
    color: ${props => props.color};
  }
`

const AuthorInfo = styled.span`
  margin-right: ${space(2)}px;
`

const TwitterHandle = styled.a`
  text-decoration: none;
  white-space: nowrap;

  ${Icon} {
    vertical-align: middle;
    margin: 0;
  }
`
