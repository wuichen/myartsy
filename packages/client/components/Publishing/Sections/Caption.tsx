import { garamond, unica } from "Assets/Fonts"
import { pMedia } from "Components/Helpers"
import { ArticleLayout, SectionLayout } from "Components/Publishing/Typings"
import React from "react"
import styled from "styled-components"

interface CaptionProps {
  caption: string
  color?: string
  index?: any
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
}

interface FigcaptionProps {
  color?: string
  layout: ArticleLayout
  sectionLayout?: SectionLayout
}

export const Caption: React.SFC<CaptionProps> = props => {
  const { caption, children, color, layout, sectionLayout } = props

  const child = children || (
    <div dangerouslySetInnerHTML={{ __html: caption }} />
  )

  return (
    <CaptionContainer>
      <Figcaption layout={layout} sectionLayout={sectionLayout} color={color}>
        {child}
      </Figcaption>
    </CaptionContainer>
  )
}

export const CaptionContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 10px 0 10px 0;

  ${pMedia.xs`
    padding: 0px 10px;
  `};
`

// includes draft placeholder class for editable text in Writer
const Figcaption = styled.div<FigcaptionProps>`
  padding: ${props => (props.sectionLayout === "fillwidth" ? "0 10px;" : "0;")}
  width: 100%;
  word-break: break-word;

  & > p, p,
  .public-DraftEditorPlaceholder-root,
  .public-DraftStyleDefault-block {
    ${props => (props.layout === "classic" ? garamond("s15") : unica("s14"))}
    color: ${props =>
      props.color ? props.color : props.layout === "classic" ? "#666" : "#999"};
    margin: 0;
    display: inline;    
  }

  & > a, a {
    color: #999;
    &:hover {
      color: black;
    }
  }

  ${pMedia.xs`
    padding: 0px;
  `}
`
