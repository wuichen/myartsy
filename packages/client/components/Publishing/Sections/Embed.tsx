import React from "react"
import styled, { StyledFunction } from "styled-components"
import { pMedia } from "../../Helpers"

interface EmbedProps {
  section: any
}

export const Embed: React.SFC<EmbedProps> = props => {
  const { url, height, mobile_height } = props.section
  return (
    <IFrame
      src={url}
      scrolling="no"
      frameBorder="0"
      height={height}
      mobileHeight={mobile_height}
    />
  )
}

interface FrameProps extends React.HTMLProps<HTMLIFrameElement> {
  mobileHeight?: number
  height: number
}

const iframe: StyledFunction<FrameProps> = styled.iframe

const IFrame = iframe`
  width: 100%;
  height: ${props => props.height + "px"};
  ${props => pMedia.sm`
    height: ${props.mobileHeight}px;
    width: 100%;
  `}
`
