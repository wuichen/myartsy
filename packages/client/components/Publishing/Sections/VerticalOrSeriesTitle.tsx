import { color, Sans } from "@artsy/palette"
import React from "react"
import styled from "styled-components"

import { getEditorialHref } from "Components/Publishing/Constants"
import { ArticleData } from "Components/Publishing/Typings"

interface Props {
  article?: ArticleData
  color?: string
  prefix?: string
  vertical?: any
}

export const VerticalOrSeriesTitle: React.SFC<Props> = props => {
  const { article, prefix } = props
  const { seriesArticle } = article
  const vertical = props.vertical
    ? props.vertical
    : article.vertical && article.vertical.name

  const seriesLink =
    seriesArticle && getEditorialHref("series", seriesArticle.slug)

  return (
    <Vertical size="3" weight="medium" color={props.color}>
      {prefix}
      {seriesArticle ? (
        <a href={seriesLink}>{seriesArticle.title}</a>
      ) : (
        <span>{vertical}</span>
      )}
    </Vertical>
  )
}

export const Vertical = styled(Sans)<{ color: string }>`
  a {
    color: ${props => props.color};
    text-decoration: none;
  }
`

VerticalOrSeriesTitle.defaultProps = {
  color: color("black100"),
}
