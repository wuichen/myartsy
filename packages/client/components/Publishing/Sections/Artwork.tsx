import React from "react"
import styled from "styled-components"
import { resize } from "../../../Utils/resizer"
import { GLOBAL_IMAGE_QUALITY } from "../Constants"
import { ArticleLayout, SectionLayout } from "../Typings"
import { ArtworkCaption } from "./ArtworkCaption"
import { ImageWrapper } from "./ImageWrapper"

export interface ArtworkProps {
  artwork: any
  color?: string
  editing?: boolean
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
  linked?: boolean
  width?: string | number
  height?: string | number
  slideshowIndex?: number
}

export class Artwork extends React.PureComponent<ArtworkProps> {
  static defaultProps = {
    linked: true,
    width: "100%",
    height: "auto",
  }

  render() {
    const {
      artwork,
      children,
      editing,
      linked,
      height,
      width,
      layout,
      slideshowIndex,
    } = this.props
    const src = resize(artwork.image, {
      width: 1200,
      quality: GLOBAL_IMAGE_QUALITY,
    })

    const Image = () => (
      <ImageWrapper
        editing={editing}
        layout={layout}
        linked={linked}
        src={src}
        width={width}
        height={height}
        alt={artwork.title}
        slideshowIndex={slideshowIndex}
      />
    )

    return (
      <ArtworkContainer>
        {linked ? (
          <ArtworkImageLink href={`/artwork/${artwork.slug}`}>
            <Image />
          </ArtworkImageLink>
        ) : (
          <Image />
        )}

        <ArtworkCaption {...this.props} />
        {children}
      </ArtworkContainer>
    )
  }
}

const ArtworkImageLink = styled.a`
  text-decoration: none;
`

const ArtworkContainer = styled.div`
  display: flex;
  flex-direction: column;
`
