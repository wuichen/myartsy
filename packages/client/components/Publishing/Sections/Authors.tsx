import React from "react"
import styled from "styled-components"
import { Author } from "./Author"

interface AuthorsProps {
  authors: object[]
  color?: string
}

export const Authors: React.SFC<AuthorsProps> = props => {
  const { authors, color } = props
  return (
    <AuthorsContainer>
      {authors.map((author, i) => (
        <Author author={author} key={i} color={color} />
      ))}
    </AuthorsContainer>
  )
}
const AuthorsContainer = styled.div`
  display: flex;
  flex-direction: column;
`
