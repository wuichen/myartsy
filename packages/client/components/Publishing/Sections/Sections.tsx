import { Box } from "@artsy/palette"
import { targetingData } from "Components/Publishing/Display/DisplayTargeting"
import { getSlideshowImagesFromArticle } from "Components/Publishing/Layouts/ArticleWithFullScreen"
import { AdDimension, AdUnit } from "Components/Publishing/Typings"
import { clone, compact, findLastIndex, get, once } from "lodash"
import React, { Component } from "react"
import ReactDOM from "react-dom"
import styled from "styled-components"
import { pMedia } from "../../Helpers"
import { DisplayAd } from "../Display/DisplayAd"
import { ArticleData } from "../Typings"
import { Authors } from "./Authors"
import { Embed } from "./Embed"
import { ImageCollection } from "./ImageCollection"
import { ImageSetPreview } from "./ImageSetPreview"
import { SectionContainer } from "./SectionContainer"
import { SocialEmbed } from "./SocialEmbed"
import { Text } from "./Text"
import { Video } from "./Video"

interface Props {
  DisplayPanel?: any
  article: ArticleData
  color?: string
  isMobile?: boolean
  showTooltips?: boolean
  isSponsored?: boolean
  isSuper?: boolean
  customWidth?: number
  isTruncatedAt?: number
  hideAds?: boolean
}

interface State {
  shouldInjectMobileDisplay: boolean
}

/**
 * When isMobile, hide sidebar and inject DisplayAd into the body of the
 * article at a specific paragraph index.
 */
const MOBILE_DISPLAY_INJECT_INDEX = 1
const MOBILE_DISPLAY_INJECT_ID_PREFIX = "__mobile_display_inject__"

export class Sections extends Component<Props, State> {
  static defaultProps = {
    isMobile: false,
  }

  displayInjectId: string

  state = {
    shouldInjectMobileDisplay: false,
  }

  UNSAFE_componentWillMount() {
    const {
      article: { layout },
      isMobile,
    } = this.props

    this.injectDisplayPanelMarker = once(this.injectDisplayPanelMarker)
    const shouldInjectMobileDisplay = isMobile && layout !== "feature"

    this.setState({
      shouldInjectMobileDisplay,
    })
  }

  componentDidMount() {
    if (this.state.shouldInjectMobileDisplay) {
      this.mountDisplayToMarker()
    }
  }

  componentDidUpdate(prevProps) {
    const { isMobile } = this.props

    if (prevProps.isMobile !== isMobile) {
      this.setState(
        {
          shouldInjectMobileDisplay: isMobile,
        },
        () => {
          if (isMobile && this.state.shouldInjectMobileDisplay) {
            this.mountDisplayToMarker()
          }
        }
      )
    }
  }

  /**
   * Inject DisplayAd after a specific paragraph index
   */
  injectDisplayPanelMarker(body) {
    const articleId = this.props.article.id
    const tag = "</p>"
    const updatedBody = compact(body.split(tag))
      .map(p => p + tag)
      .reduce((arr, block, paragraphIndex) => {
        if (paragraphIndex === MOBILE_DISPLAY_INJECT_INDEX) {
          this.displayInjectId = `${MOBILE_DISPLAY_INJECT_ID_PREFIX}${articleId}`
          return arr.concat([block, `<div id="${this.displayInjectId}"></div>`])
        } else {
          return arr.concat([block])
        }
      }, [])
      .join("")

    return updatedBody
  }

  mountDisplayToMarker() {
    const { DisplayPanel } = this.props
    const displayMountPoint = document.getElementById(this.displayInjectId)

    if (displayMountPoint) {
      ReactDOM.render(<DisplayPanel />, displayMountPoint)
    } else {
      console.error(
        "(reaction/Sections.tsx) Error mounting Display: DOM node ",
        "not found",
        displayMountPoint
      )
    }
  }

  getContentStartIndex = () => {
    const {
      article: { layout, sections },
    } = this.props

    if (layout === "feature") {
      const firstText = sections.findIndex(section => {
        return section.type === "text"
      })
      return firstText
    }
  }

  getContentEndIndex = () => {
    const {
      article: { layout, sections },
    } = this.props

    if (["feature", "standard"].includes(layout)) {
      const lastText = findLastIndex(sections, section => {
        return section.type === "text"
      })
      return lastText
    }
  }

  getSection(section, index) {
    const { article, color, customWidth, isMobile, showTooltips } = this.props
    const targetHeight =
      customWidth && (customWidth > 750 && !isMobile) ? 750 : 500
    const size = customWidth && { width: customWidth }
    const fullscreenImages = getSlideshowImagesFromArticle(article)

    const sections = {
      image_collection: (
        <ImageCollection
          color={color}
          sectionLayout={section.layout}
          articleLayout={article.layout}
          images={section.images}
          targetHeight={targetHeight}
          gutter={10}
          size={size}
          fullscreenImages={fullscreenImages}
        />
      ),
      image_set: (
        <ImageSetPreview
          section={section}
          color={color}
          fullscreenImages={fullscreenImages}
        />
      ),
      video: <Video section={section} color={color} />,
      embed: <Embed section={section} />,
      social_embed: <SocialEmbed section={section} />,
      text: (
        <Text
          color={color}
          html={section.body}
          layout={article.layout}
          isContentStart={index === this.getContentStartIndex()}
          isContentEnd={index === this.getContentEndIndex()}
          showTooltips={showTooltips}
        />
      ),
      default: false,
    }

    const sectionComponent = sections[section.type] || sections.default
    return sectionComponent
  }

  getAdUnit(index: number, indexAtFirstAd: number) {
    const { isMobile } = this.props

    if (index === indexAtFirstAd) {
      return isMobile
        ? AdUnit.Mobile_Feature_InContentLeaderboard1
        : AdUnit.Desktop_Feature_Leaderboard1
    }

    if (index === 6) {
      return isMobile
        ? AdUnit.Mobile_Feature_InContentLeaderboard2
        : AdUnit.Desktop_Feature_Leaderboard2
    }

    return isMobile
      ? AdUnit.Mobile_Feature_InContentLeaderboard3
      : AdUnit.Desktop_Feature_LeaderboardRepeat
  }

  /**
   * Add a margin top when to the ad when the section is any of the following:
   * 1) Text section without <br> tags wrapping the <p> or <h2> content
   * 3) Image set section
   * 4) Image collection with a caption/title
   */
  getAdMarginTop(section) {
    const shouldAddTopMargin = () => {
      switch (section.type) {
        case "image_set":
          return true
        case "image_collection":
          get(section, ["images", "[0]", "caption"])
          return true
        case "text":
          section.body.includes("<br>")
          return true
        default:
          return false
      }
    }

    return shouldAddTopMargin() ? "calc(100% - 96%)" : null
  }

  renderSections() {
    const {
      article,
      customWidth,
      isMobile,
      isSponsored,
      isSuper,
      isTruncatedAt,
      hideAds,
    } = this.props
    const { shouldInjectMobileDisplay } = this.state
    let quantityOfAdsRendered = 0
    let firstAdInjected = false
    let placementCount = 1
    let displayMarkerInjected = false
    let indexAtFirstAd = null

    const articleSections = isTruncatedAt
      ? clone(article.sections).slice(0, isTruncatedAt)
      : article.sections

    const renderedSections = articleSections.map((sectionItem, index) => {
      let section = sectionItem
      let ad = null
      const shouldInject =
        shouldInjectMobileDisplay &&
        sectionItem.type === "text" &&
        !displayMarkerInjected

      /**
       *  Possible data types for sectionItem.type are
       *  callout, embed, social_embed, text, slideshow, image_set, and image_collection.
       *  Here we want to inject the first ad after the first image_collection OR image_set data type.
       * */
      let shouldInjectNewAds =
        article.layout === "feature" &&
        (sectionItem.type === "image_collection" ||
          sectionItem.type === "image_set") &&
        !firstAdInjected &&
        !isSuper &&
        !hideAds

      if (firstAdInjected) {
        placementCount++
      }

      if (placementCount % 6 === 0 && quantityOfAdsRendered < 2) {
        // only render 2 ads on Features
        shouldInjectNewAds = true
      }

      if (shouldInjectNewAds) {
        quantityOfAdsRendered++

        const adDimension = isMobile
          ? isSponsored
            ? AdDimension.Mobile_Sponsored_Feature_InContentLeaderboard1
            : AdDimension.Mobile_Feature_InContentLeaderboard1
          : AdDimension.Desktop_NewsLanding_Leaderboard1

        const marginTop = this.getAdMarginTop(section)

        firstAdInjected = true
        indexAtFirstAd = indexAtFirstAd === null ? index : indexAtFirstAd // only set this value once; after the index where 1st ad injection is found

        ad = (
          <AdWrapper mt={marginTop}>
            <DisplayAd
              adUnit={this.getAdUnit(placementCount, indexAtFirstAd)}
              adDimension={adDimension}
              targetingData={targetingData(
                article,
                isSponsored ? "sponsorfeature" : "feature"
              )}
            />
          </AdWrapper>
        )
      }

      if (shouldInject) {
        try {
          section = clone(sectionItem)
          section.body = this.injectDisplayPanelMarker(section.body)
          displayMarkerInjected = true
        } catch (error) {
          console.error(
            "(reaction/Sections.jsx) Error injecting Display:",
            error
          )
        }
      }

      const child = this.getSection(section, index)

      if (child) {
        return (
          <SectionContainer
            key={index}
            articleLayout={article.layout}
            section={section}
            customWidth={customWidth}
          >
            {child}
            {ad}
          </SectionContainer>
        )
      }
    })

    return renderedSections
  }

  renderAuthors() {
    const {
      color,
      article: { authors },
    } = this.props

    if (authors) {
      return (
        <SectionContainer>
          <Authors authors={authors} color={color} />
        </SectionContainer>
      )
    }
  }

  renderPostScript() {
    const { article, color } = this.props
    const { layout, postscript } = article

    if (postscript) {
      return (
        <SectionContainer>
          <Text
            color={color}
            html={postscript}
            layout={layout}
            postscript={Boolean(postscript)}
          />
        </SectionContainer>
      )
    }
  }

  render() {
    const { article } = this.props

    return (
      <StyledSections layout={article.layout}>
        {this.renderSections()}
        {this.renderAuthors()}
        {this.renderPostScript()}
      </StyledSections>
    )
  }
}

const chooseMargin = layout => {
  if (layout === "standard") {
    return "0;"
  } else if (layout === "feature") {
    return "80px auto 0 auto;"
  }
}

export const StyledSections = styled.div<{ layout: string }>`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  margin: ${({ layout }) => chooseMargin(layout)};
  max-width: ${({ layout }) => (layout === "standard" ? "780px" : "auto")};

  ${({ layout }) => pMedia.xl`
    max-width: ${layout === "standard" ? "680px" : "auto"};
    ${layout === "feature" ? "margin: 80px auto 0 auto" : ""}
  `}

  ${({ layout }) => pMedia.md`
    max-width: ${layout === "standard" ? "780px" : "auto"};
  `}

  ${({ layout }) => pMedia.xs`
    max-width: ${layout === "standard" ? "780px" : "auto"};
    ${layout === "feature" ? "margin: 30px auto 0 auto" : ""}
  `}
`
const AdWrapper = styled(Box)`
  width: 100vw;
  margin-left: calc(-50vw + 50%);
`
