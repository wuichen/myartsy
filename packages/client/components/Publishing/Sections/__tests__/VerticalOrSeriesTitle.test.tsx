import { VerticalOrSeriesTitle } from "Components/Publishing/Sections/VerticalOrSeriesTitle"
import { mount } from "enzyme"
import "jest-styled-components"
import { clone } from "lodash"
import React from "react"
import renderer from "react-test-renderer"

import {
  FeatureArticle,
  SeriesArticleSponsored,
  StandardArticle,
} from "Components/Publishing/Fixtures/Articles"

it("renders properly for a feature article", () => {
  const component = renderer
    .create(<VerticalOrSeriesTitle article={FeatureArticle} />)
    .toJSON()
  expect(component).toMatchSnapshot()
})

it("renders properly for a standard article", () => {
  const component = renderer
    .create(<VerticalOrSeriesTitle article={FeatureArticle} />)
    .toJSON()
  expect(component).toMatchSnapshot()
})

it("renders the article vertical", () => {
  const component = mount(<VerticalOrSeriesTitle article={FeatureArticle} />)
  expect(component.text()).toMatch(FeatureArticle.vertical.name)
})

it("renders a series if feature and fullscreen", () => {
  const FeatureSeriesArticle = clone({
    ...FeatureArticle,
    seriesArticle: SeriesArticleSponsored,
  })
  const component = mount(
    <VerticalOrSeriesTitle article={FeatureSeriesArticle} />
  )
  expect(component.text()).toMatch(SeriesArticleSponsored.title)
})

it("renders with props vertical", () => {
  const component = mount(
    <VerticalOrSeriesTitle
      article={StandardArticle}
      vertical="Missing Vertical"
    />
  )
  expect(component.text()).toMatch("Missing Vertical")
})

it("renders with props prefix", () => {
  const component = mount(
    <VerticalOrSeriesTitle article={StandardArticle} prefix="More in " />
  )
  expect(component.text()).toMatch("More in Art Market")
})
