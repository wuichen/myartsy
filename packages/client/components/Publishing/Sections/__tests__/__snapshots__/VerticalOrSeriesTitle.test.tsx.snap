// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly for a feature article 1`] = `
.c1 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  color: #000;
}

.c0 a {
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0 c1"
  color="#000"
  fontFamily={
    Object {
      "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
      "fontWeight": 500,
    }
  }
  fontSize="14px"
>
  <span>
    Creativity
  </span>
</div>
`;

exports[`renders properly for a standard article 1`] = `
.c1 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  color: #000;
}

.c0 a {
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0 c1"
  color="#000"
  fontFamily={
    Object {
      "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
      "fontWeight": 500,
    }
  }
  fontSize="14px"
>
  <span>
    Creativity
  </span>
</div>
`;
